server/src/main/scala/controllers/application/ShiftsController.scala

```diff
@@ -165,23 +159,69 @@ class ShiftsController @Inject()(cc: ControllerComponents,
+            Future.successful(BadRequest("Can only add one shift at a time when existing shifts are present"))
+          else
+            addAShift(newShifts.head)
+        } else
+          createShifts(newShifts)
+      }.recover { case e: Exception =>
+        log.error(s"Failed to process shifts: ", e.printStackTrace())
+        BadRequest(s"Failed to process shifts: ${e.getMessage}")
+      }
+    }.getOrElse(Future.successful(BadRequest("Expecting JSON data")))
+  }
+
+  private def addAShift(newShift: Shift) = {
+    ctrl.shiftsService.getShift(newShift.port, newShift.terminal, newShift.shiftName, newShift.startDate).flatMap {
+      case Some(_) =>
+        Future.successful(BadRequest(s"Shift with name ${newShift.shiftName} already exists on ${newShift.startDate}"))
+      case None =>
+        ctrl.shiftsService.saveShift(Seq(newShift)).flatMap { _ =>
           shiftAssignmentsService.allShiftAssignments.flatMap { allShiftAssignments =>
-            val updatedAssignments = updateWithShiftDefaultStaff(shifts, allShiftAssignments)
-            shiftAssignmentsService.updateShiftAssignments(updatedAssignments).map { s =>
-              ctrl.shiftStaffRollingService.upsertShiftStaffRolling(shiftStaffRollingCreating(shifts))
-              Ok(write(s))
-            }
+            ctrl.shiftsService.getOverlappingStaffShifts(newShift.port, newShift.terminal, newShift)
+              .flatMap { overridingShifts =>
+                val updatedAssignments = StaffingUtil.addAssignments(
+                  newShift = newShift,
+                  overridingShift = overridingShifts,
+                  allShifts = allShiftAssignments)
+                shiftAssignmentsService.updateShiftAssignments(updatedAssignments).map { s =>
+                  ctrl.shiftStaffRollingService.upsertShiftStaffRolling(shiftStaffingRollingWhileEditing(newShift))
+                  Ok(write(s))
+                }.recoverWith {
+                  case e: Exception =>
+                    log.error(s"Failed to add new shift with assignments: ", e.printStackTrace())
+                    Future.successful(BadRequest(s"Failed to update shifts: ${e.getMessage}"))
+                }
+              }
           }.recoverWith {
-            case e: Exception => Future.successful(BadRequest(s"Failed to update shifts: ${e.getMessage}"))
+            case e: Exception =>
+              log.error(s"Failed to get all shifts assignments: ", e.printStackTrace())
+              Future.successful(BadRequest(s"Failed to update shifts: ${e.getMessage}"))
           }
+        }.recoverWith {
+          case e: Exception =>
+            log.error(s"Failed to create new shift: ", e.printStackTrace())
+            Future.successful(BadRequest(s"Failed to create new shift: ${e.getMessage}"))
         }
-      } catch {
-        case _: DeserializationException => Future.successful(BadRequest("Invalid shift format"))
+    }
+  }
+
+  private def createShifts(shifts: Seq[Shift]) = {
```

Comment:
I still don't see a need for this function. `addAShift` should work the same whether we're adding the first ever shift or an additional shift